pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,283 @@
+          tableScan, partitionFunction);
+      return partitionFunction;
+    } else {
+      return tablePartitionInfo.getPartitionFunctionName();
+    }
+  }
+
+  /**
+   * Get the partition parallelism from the hint, if any, otherwise use the default value of 1.
+   */
+  private static int getPartitionParallelism(LogicalTableScan tableScan, Map<String, String> kvOptions) {
+    int partitionParallelism = 1;
+    String partitionParallelismStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM);
+    if (partitionParallelismStr != null) {
+      try {
+        int explicitPartitionParallelism = Integer.parseInt(partitionParallelismStr);
+        if (explicitPartitionParallelism != partitionParallelism) {
+          LOGGER.debug("Override implicit table hint for {} with explicit partition parallelism: {}", tableScan,
+              explicitPartitionParallelism);
+          partitionParallelism = explicitPartitionParallelism;
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            "Invalid partition parallelism: " + partitionParallelismStr + " for table: " + tableScan);
+      }
+    }
+    return partitionParallelism;
+  }
+
+  /**
+   * Get the partition size from the hint, if any, otherwise use the partition size from the table partition info.
+   */
+  private static int getPartitionSize(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    int partitionSize = tablePartitionInfo.getNumPartitions();
+    String partitionSizeStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+    if (partitionSizeStr != null) {
+      try {
+        int explicitPartitionSize = Integer.parseInt(partitionSizeStr);
+        if (explicitPartitionSize != partitionSize) {
+          LOGGER.debug("Override implicit table hint for {} with explicit partition size: {}", tableScan,
+              explicitPartitionSize);
+          partitionSize = explicitPartitionSize;
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("Invalid partition size: " + partitionSizeStr + " for table: " + tableScan);
+      }
+    }
+    return partitionSize;
+  }
+
+  /**
+   * Get the table partition info for the given table scan, transforming Calcite's table name to Pinot's table name
+   * and adding the table type if needed.
+   */
+  @Nullable
+  private TablePartitionInfo getTablePartitionInfo(LogicalTableScan tableScan) {
+    List<String> qualifiedName = tableScan.getTable().getQualifiedName();
+    String tableName;
+    if (qualifiedName.size() == 1) {
+      tableName = qualifiedName.get(0);
+    } else {
+      tableName = DatabaseUtils.translateTableName(qualifiedName.get(1), qualifiedName.get(0));
+    }
+
+    if (TableNameBuilder.getTableTypeFromTableName(tableName) == null) {
+      tableName = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
```

Comment:
(MAJOR) This is not robust, and can cause wrong behavior on real-time or hybrid table. We'll need to check both real-time and offline and apply hint when they match each other, or only one exists

Reply:
How complicated is this check? Are we going to replicate a lot of logic within `WorkerManager`?
I was thinking only apply the hint when both tables have the same `partitionColumn`, `partitionFunctionName` and `numPartitions`. Then I realized another issue: we need to know if a table exists to figure out whether it is either unpartitioned, or doesn't exist.

Reply:
>  we need to know if a table exists

What do you mean? If the table doesn't exits we would have failed at validation time, right? Do you mean to actually have segments?

My concern here is that `WorkerManager` fails if there is at least one partition that is not fully replicated (see https://github.com/apache/pinot/blob/49126f28356a4fe61c7a23d330555109b4759ba0/pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java#L450). Therefore if we just compare `partitionColumn`, `partitionFunctionName` and `numPartitions` and the table is not fully replicated, we will end up failing at `WorkerManager`.

My suggestion is to modify `WorkerManager` to provide a method similar to `getPartitionTableInfo` that we can call from the rule. This method should check whether partition info is equal and if for each partition there is at least one server with a full copy of the partition. The main differences with current `getPartitionTableInfo` is that:
- It cannot return `PartitionTableInfo`, given it is an internal class the rule don't care about.
- It cannot call `getTablePartitionInfo`, which doesn't only _get_ the info but also compares values on the hint with what is stored in the routing manager.

Reply:
I've changed the code. Can you take a look to the new implementation that delegates on worker manager?